opencog/query/PatternMatch.cc

```diff
@@ -378,13 +379,29 @@ bool PatternLink::satisfy(PatternMatchCallback& pmcb) const
 	{
 		dbgprt("BEGIN COMPONENT GROUNDING %zu of %zu: ======================\n",
 		       i + 1, _num_comps);
+
+		Pattern pat = PatternLinkCast(_component_patterns.at(i))->get_pattern();
+		bool is_pure_optional = false;
+		if (pat.mandatory.size() == 0 and pat.optionals.size() > 0)
+			is_pure_optional = true;
+
 		// Pass through the callbacks, collect up answers.
 		PMCGroundings gcb(pmcb);
 		PatternLinkPtr clp(PatternLinkCast(_component_patterns.at(i)));
 		clp->satisfy(gcb);
 
-		comp_var_gnds.push_back(gcb._var_groundings);
-		comp_term_gnds.push_back(gcb._term_groundings);
+		// Special handling for disconnected pure optionals -- Returns false to
+		// end the search if this disconnected pure optional is found
+		if (is_pure_optional)
+		{
+			DefaultPatternMatchCB* dpmcb = dynamic_cast<DefaultPatternMatchCB*>(&pmcb);
+			if (dpmcb->optionals_present()) return false;
```

Comment:
Misegana is right, This is guaranteed to crash if the callback is not the default callback! (the case will result in dpmcb in being null) ... some of the custom callbacks might not know what an "optional" is (maybe e.g. the fuzzy matcher??)

The reason that this is safe:
https://github.com/leungmanhin/atomspace/blob/master/opencog/query/Implicator.cc#L117-L120
is because, over there, we know, with 100% certainty that the default matcher is being used, because its all internal to that file; its just a utility. Ther is no way a non-default db could ever end up there.